File: src/main/java/net/igneo/icv/init/TrigMath.java
package net.igneo.icv.init;

public class TrigMath {
    private static final double SECOND_ORDER_COEFFICIENT = 0.00000000109111226653103698;
    
    private static double quasiCos2(double x) {
        return 1.0d - SECOND_ORDER_COEFFICIENT * x * x;
    }
    
    public static double[] sincos(int intAngle) {
        int shifter = (intAngle ^ (intAngle << 1)) & 0xC000;
        double x = ((intAngle + shifter) << 17) >> 16;
        
        double cosx = quasiCos2(x);
        double sinx = Math.sqrt(1.0d - cosx * cosx);
        
        if ((shifter & 0x4000) != 0) {
            double temp = cosx;
            cosx = sinx;
            sinx = temp;
        }
        
        if (intAngle < 0) {
            sinx = -sinx;
        }
        
        if ((shifter & 0x8000) != 0) {
            cosx = -cosx;
        }
        
        return new double[]{sinx, cosx};
    }
    
    public static double tan(int intAngle) {
        double[] result = sincos(intAngle);
        double sin = result[0];
        double cos = result[1];
        
        if (Math.abs(cos) < 1e-10) {
            return sin > 0 ? Double.POSITIVE_INFINITY : Double.NEGATIVE_INFINITY;
        }
        
        return sin / cos;
    }
    
    public static int radiansToInternalAngle(double radians) {
        return (int) ((radians % (2 * Math.PI)) * 32768 / (2 * Math.PI)) & 0xFFFF;
    }
    
    
}

File: src/main/java/net/igneo/icv/init/TrigMath.kt
package net.igneo.icv.init

import kotlin.math.abs
import kotlin.math.sqrt

object TrigMath {
    private const val SECOND_ORDER_COEFFICIENT = 1.0911122665310369E-9

    private fun quasiCos2(x: Double): Double {
        return 1.0 - SECOND_ORDER_COEFFICIENT * x * x
    }

    fun sincos(intAngle: Int): DoubleArray {
        val shifter = (intAngle xor (intAngle shl 1)) and 0xC000
        val x = (((intAngle + shifter) shl 17) shr 16).toDouble()

        var cosx = quasiCos2(x)
        var sinx = sqrt(1.0 - cosx * cosx)

        if ((shifter and 0x4000) != 0) {
            val temp = cosx
            cosx = sinx
            sinx = temp
        }

        if (intAngle < 0) {
            sinx = -sinx
        }

        if ((shifter and 0x8000) != 0) {
            cosx = -cosx
        }

        return doubleArrayOf(sinx, cosx)
    }

    fun tan(intAngle: Int): Double {
        val result = sincos(intAngle)
        val sin = result[0]
        val cos = result[1]

        if (abs(cos) < 1e-10) {
            return if (sin > 0) Double.POSITIVE_INFINITY else Double.NEGATIVE_INFINITY
        }

        return sin / cos
    }

    fun radiansToInternalAngle(radians: Double): Int {
        return ((radians % (2 * Math.PI)) * 32768 / (2 * Math.PI)).toInt() and 0xFFFF
    }
}